ceph/ceph.go

```diff
@@ -92,83 +104,174 @@ func New() *Ceph {
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Ceph perf schema command execution failed for socket %+v, err=%+v\n",
+				filepath.Join(ceph.socket.path, daemon.fullName), err)
+			return err
+		}
+		perfSchemaData := make(map[string]interface{})
+		if err := json.Unmarshal(out, &perfSchemaData); err != nil {
+			fmt.Fprintf(os.Stderr, "Cannot unmarshal JSON object from ceph-daemon socket schema, err=%+v\n", err)
+			return err
+		}
+		ceph.getJSONPathsByNamespace(perfSchemaData, jsonSchema, []string{}, []string{}, ceph.schema[daemon.name])
 	}
 
-	daemonNameSplit := strings.Split(trimPrefixAndSuffix(daemon, ceph.socket.prefix, "."+ceph.socket.ext), ".")
-	daemonName := daemonNameSplit[0]
-	daemonID := daemonNameSplit[1]
+	return nil
+}
+
+// getDaemon returns daemon using name and id
+func (ceph *Ceph) getDaemon(name, id string) *Daemon {
+	for _, d := range ceph.daemons {
+		if d.name == name && d.id == id {
+			return &d
+		}
+	}
+	return nil
+}
 
-	metrics := []plugin.MetricType{}
+// getCephDaemonMetrics executes "ceph --admin-daemon perf dump" command for defined daemon-socket and returns metrics matching with mts
+// If provided metric list is nil, function returns all available Ceph metrics using "ceph --admin-daemon perf schema" command
+// Delta metrics (sum / avgcount) are calculated automatically
+func (ceph *Ceph) getCephDaemonMetrics(mts []plugin.Metric, daemon Daemon) ([]plugin.Metric, error) {
+	var perfDumpData map[string]interface{}
+	metricsCollected := []plugin.Metric{}
+
+	// Go throught schema and build metric list
+	for schemaNs, schemaData := range ceph.schema[daemon.name] {
+		description := ""
+		if schemaData.(map[string]interface{})["description"] != nil {
+			description = schemaData.(map[string]interface{})["description"].(string)
+		}
 
-	for _, m := range mts {
-		// Get metrics defined in task for this daemon
-		if matchSlice(m.Namespace().Strings()[daemonNameIndex:daemonIDIndex+1], daemonNameSplit) {
-			daemonMetrics := make(map[string]interface{})
-			ceph.getJSONDataByNamespace(jsonData, m.Namespace().Strings()[daemonIDIndex+1:], []string{}, daemonMetrics)
+		metricMeta := plugin.Metric{
+			Namespace:   plugin.NewNamespace(nsVendor, nsClass, nsType, daemon.name).AddDynamicElement(daemon.name+"_id", fmt.Sprintf("ID of %s daemon", daemon.name)).AddStaticElements(strings.Split(schemaNs, "/")...),
+			Tags:        map[string]string{"daemon_source": fmt.Sprintf("%s.%s", daemon.name, daemon.id)},
+			Timestamp:   time.Now(),
+			Description: description,
+		}
 
-			// No metrics found for desired namespace
-			if len(daemonMetrics) == 0 {
-				daemonMetrics[strings.Join(m.Namespace().Strings()[daemonIDIndex+1:], "/")] = nil
+		// Detect metric unit
+		if schemaData.(map[string]interface{})["type"] != nil {
+			if checkBit(uint8(schemaData.(map[string]interface{})["type"].(float64)), cephTypeFloat) {
+				metricMeta.Unit = "float"
```

Comment:
I removed unit, also I don't use variable to store type info, just check type right before data type conversion (plugin converts data to uint64 if needed, default float64 or nil otherwise):

```
// Return complete metric with data converted to desired type
				metricMeta.Namespace[4].Value = daemon.id
				if check, err := isCephTypeUint64(schemaData); err != nil {
					return nil, err
				} else if check && data != nil {
					if val, ok := data.(float64); ok {
						metricMeta.Data = uint64(val)
					}
				} else {
					metricMeta.Data = data
				}
				metricsCollected = append(metricsCollected, metricMeta)
```